server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java

```diff
@@ -142,6 +150,62 @@ public void testCancelException() {
         assertPropagatesInterrupt(() -> future.actionGet(10, TimeUnit.SECONDS));
     }
 
+    public void testAssertCompleteAllowedAllowsConcurrentCompletes() throws InterruptedException {
+        assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
+        try (TestThreadPool tp = new TestThreadPool(getTestName())) {
+            final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
+            final int threads = 4;
+            final AtomicBoolean running = new AtomicBoolean(true);
+            final CountDownLatch startBarrier = new CountDownLatch(threads + 1);
+            final CompletionService<?> cs = new ExecutorCompletionService<>(tp.generic());
+            final List<Future<?>> futures = new ArrayList<>();
+            for (int i = 0; i < threads; i++) {
+                futures.add(cs.submit(new FutureCompleter(i, startBarrier, futureReference, running), null));
+            }
+            startBarrier.countDown();
+            safeAwait(startBarrier);
+            cs.poll(250, TimeUnit.MILLISECONDS);
```

Comment:
This test will unfortunately run the full 250 in the happy case, on my desktop it seemed to fail around the 80ms mark fairly reliably before the fix.

Reply:
I'd rather a test which fails _sometimes_ but always runs quickly, because we run this so often in CI that it'll catch the bug pretty soon anyway.

Reply:
Cool, I reduced to 20ms which is still surprisingly reliable.

Reply:
I'm wondering why we need such a complicated test to reproduce this. Something like this seems to work too:

```diff
diff --git a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
index 2ca914eb23c..68c706c7f3c 100644
--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -14,6 +14,8 @@ import org.elasticsearch.common.util.concurrent.UncategorizedExecutionException;
 import org.elasticsearch.core.Assertions;
 import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.RemoteTransportException;

 import java.util.concurrent.CancellationException;
@@ -180,4 +182,19 @@ public class PlainActionFutureTests extends ESTestCase {
             }
         };
     }
+
+    public void testConcurrentCompletion() {
+        try (var threadPool = new TestThreadPool(getTestName())) {
+            final var future = new PlainActionFuture<>();
+            final var threadCount = threadPool.info(ThreadPool.Names.GENERIC).getMax();
+            final var barrier = new CyclicBarrier(threadCount);
+            for (int taskIndex = 0; taskIndex < threadCount; taskIndex++) {
+                threadPool.generic().execute(() -> {
+                    safeAwait(barrier);
+                    future.onResponse(null);
+                });
+            }
+            assertNull(safeGet(future));
+        }
+    }
 }
```

Reply:
slightly nicer:

```diff
diff --git a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
index 2ca914eb23c..4893dfc91e8 100644
--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -9,11 +9,14 @@
 package org.elasticsearch.action.support;

 import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.ActionRunnable;
 import org.elasticsearch.common.util.concurrent.FutureUtils;
 import org.elasticsearch.common.util.concurrent.UncategorizedExecutionException;
 import org.elasticsearch.core.Assertions;
 import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.RemoteTransportException;

 import java.util.concurrent.CancellationException;
@@ -180,4 +183,18 @@ public class PlainActionFutureTests extends ESTestCase {
             }
         };
     }
+
+    public void testConcurrentCompletion() {
+        try (var threadPool = new TestThreadPool(getTestName())) {
+            final var future = new PlainActionFuture<>();
+            final var executorName = randomFrom(ThreadPool.Names.GENERIC, ThreadPool.Names.MANAGEMENT);
+            final var threadCount = threadPool.info(executorName).getMax();
+            final var executor = threadPool.executor(executorName);
+            final var barrier = new CyclicBarrier(threadCount);
+            for (int i = 0; i < threadCount; i++) {
+                executor.execute(ActionRunnable.run(future, () -> safeAwait(barrier)));
+            }
+            assertNull(safeGet(future));
+        }
+    }
 }
```

Reply:
That doesn't seem to trigger it on my machine (It's been running for >5000 times without a failure so far). I will try and inline the logic so we don't need the inner class. I'm sure it can be made simpler.

Reply:
Weird, it fails pretty reliably for me. Although 5k iterations isn't very many.

Reply:
Try running with `-Dtests.iters=1000 -Dtests.failfast=true`, no need to wait for seconds+ for Gradle to start up on each iteration.

Reply:
I'm not sure what the difference is, but I can't get it to fail. This M3 MacBook seems to eat race conditions for breakfast (I also had trouble reproducing [this one](https://github.com/elastic/elasticsearch/issues/109866)).

I ran for 800,000 iterations which took ~48 minutes and produced zero failures where the latest iteration of my longer test fails almost immediately. While I accept that CI Is probably going to yield these failures more reliably, I think it's probably best if we have a test that reproduces on typical dev hardware.

Reply:
> I think it's probably best if we have a test that reproduces on typical dev hardware.

Hmm I think this is the wrong way round - if "typical dev hardware" isn't able to reproduce genuine race conditions that happen in CI/production then I don't think we should work around that by making one specific test more complex. I'd recommend using a GCP machine for long-running repro attempts rather than your laptop since it'll be closer to real production infra.

Reply:
I think my version of the test is about as slender as it can be now (thanks for the feedback), but I'm OK with replacing it with the much shorter test if you'd prefer.
I'm satisfied that the issue was there and the change fixes it, and it sounds like that test will fail in CI if we regress, so I'm happy to go with consensus on the cost/benefit of the additional complexity.